pennylane/labs/intermediate_reps/rowcol.py

```diff
@@ -0,0 +1,214 @@
+    # First, traverse the subtrees attached to the current node ``source``
+    for child in tree.neighbors(source):
+        # The graph does not know of the tree structure, so we need to make sure we skip
+        # the parent of ``source`` among its neighbors.
+        if child != source_parent:
+            # Recurse
+            out.extend(postorder_traverse(tree, child, source))
+    # Second, attach the current node itself, together with its parent
+    if source_parent is not None:
+        out.append((source, source_parent))
+
+    return out
+
+
+def preorder_traverse(tree: nx.Graph, source: int, source_parent: int = None):
+    """Pre-order traverse a tree graph, starting from (but excluding) the node ``source``.
+
+    Args:
+        tree (nx.Graph): Tree graph to traverse. Must contain ``source``. Must contain
+            ``source_parent`` if it is not None. Typing assumes integer-labeled nodes.
+        source (int): Node to start the traversal from
+        source_parent (Optional[int]): Parent node of ``source`` in ``tree``.
+
+    Returns:
+        list[tuple[int]]: Pairs of nodes that constitute pre-order traversal of ``tree``
+        starting at ``source``. Strictly speaking, the traversal is encoded in the first
+        entry of each pair, and the second entry simply is the parent node for each first entry.
+
+    A useful illustration of depth-first tree traversals can be found on
+    `Wikipedia <https://en.wikipedia.org/wiki/Tree_traversal#Depth-first%20search>`__.
+    """
+    out = []
+
+    # First, attach the current node itself, together with its parent
+    if source_parent is not None:
+        out.append((source, source_parent))
+    # Second, traverse the subtrees attached to the current node ``source``
+    for child in tree.neighbors(source):
+        # The graph does not know of the tree structure, so we need to make sure we skip
+        # the parent of ``source`` among its neighbors.
+        if child != source_parent:
+            # Recurse
+            out.extend(preorder_traverse(tree, child, source))
+
+    return out
+
+
+def _update(P: np.ndarray, cnots: list[tuple[int]], control: int, target: int):
+    """In-place apply update corresponding to a CNOT on wires ``(control, target)``
+    to parity matrix P and list of CNOT gates ``cnots``."""
+    P[target] += P[control]
+    cnots.append((control, target))
+
+
+F_2 = galois.GF(2)
+
+
+def _get_S(P: np.ndarray, idx: int, node_set: Iterable[int], mode: str):
+    # Find S (S') either by simply extracting a column or by solving a linear system for the row
+    if mode == "column":
+        b = P[:, idx]
+    else:
+        P = F_2(P)
+        e_i = F_2.Zeros(len(P))
+        e_i[idx] = 1
+        b = np.linalg.solve(P.T, e_i)  # This solve step is over F_2!
+    S = set(np.where(b)[0])
+    # Add the node ``idx`` itself
+    S.add(idx)
+    # Manually remove nodes from S that are no longer part of ``connectivity``.
+    S = S.intersection(node_set)
```

Comment:
why is this necessary?

Reply:
`S` is determined from `P` in any case. But we never remove rows/columns from `P` that belong to nodes that have been removed from `connectivity`. So `S` might contain `0`, for example, even after the `0`th row and column have been eliminated and the node `0` was removed from `connectivity`.
As a consequence, we have to remove such already-eliminated nodes from `S` manually (or start slicing `P`, which felt like a nuisance to me), and I do this here with the `node_set`, which is just a shorthand/cache to `set(connectivity.nodes())`. One could also use `set(connectivity.nodes())` directly, but then we would construct this set over and over again, instead of maintaining `node_set` and sequentially removing nodes from it :)

Reply:
Let me know whether I should update the code in some way, or document more. :)

Reply:
ahh gotcha! yeah slicing seems annoying indeed, so this makes sense :)